// skynet-js/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
exports.__esModule = true;
exports.defaultSkynetPortalUrl = exports.defaultPortalUrl = exports.uriSkynetPrefix = exports.uriHandshakeResolverPrefix = exports.uriHandshakePrefix = exports.parseSkylink = exports.MAX_REVISION = exports.getRootDirectory = exports.getRelativeFilePath = exports.genKeyPairFromSeed = exports.genKeyPairAndSeed = exports.deriveChildSeed = exports.SkynetClient = void 0;
var client_1 = require("./client");
__createBinding(exports, client_1, "SkynetClient");
var crypto_1 = require("./crypto");
__createBinding(exports, crypto_1, "deriveChildSeed");
__createBinding(exports, crypto_1, "genKeyPairAndSeed");
__createBinding(exports, crypto_1, "genKeyPairFromSeed");
var file_1 = require("./utils/file");
__createBinding(exports, file_1, "getRelativeFilePath");
__createBinding(exports, file_1, "getRootDirectory");
var number_1 = require("./utils/number");
__createBinding(exports, number_1, "MAX_REVISION");
var skylink_1 = require("./utils/skylink");
__createBinding(exports, skylink_1, "parseSkylink");
__createBinding(exports, skylink_1, "uriHandshakePrefix");
__createBinding(exports, skylink_1, "uriHandshakeResolverPrefix");
__createBinding(exports, skylink_1, "uriSkynetPrefix");
var url_1 = require("./utils/url");
__createBinding(exports, url_1, "defaultPortalUrl");
__createBinding(exports, url_1, "defaultSkynetPortalUrl");
