// skynet-js/dist/skydb.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getJSON = getJSON;
exports.setJSON = setJSON;

var _tweetnacl = require("tweetnacl");

var _number = require("./utils/number");

var _skylink = require("./utils/skylink");

var _string = require("./utils/string");

/**
 * Gets the JSON object corresponding to the publicKey and dataKey.
 *
 * @param this - SkynetClient
 * @param publicKey - The user public key.
 * @param dataKey - The key of the data to fetch for the given user.
 * @param [customOptions] - Additional settings that can optionally be set.
 * @returns - The returned JSON and revision number.
 * @throws - Will throw if the returned signature does not match the returned entry, or if the skylink in the entry is invalid.
 */
async function getJSON(publicKey, dataKey, customOptions) {
  const opts = { ...this.customOptions,
    ...customOptions
  }; // Lookup the registry entry.

  const {
    entry
  } = await this.registry.getEntry(publicKey, dataKey, opts);

  if (entry === null) {
    return {
      data: null,
      revision: null
    };
  } // Download the data in that Skylink.


  const skylink = entry.data;
  const {
    data
  } = await this.getFileContent(skylink, opts);

  if (typeof data !== "object" || data === null) {
    throw new Error("File data for the entry at data key '".concat(dataKey, "' is not JSON."));
  }

  return {
    data,
    revision: entry.revision
  };
}
/**
 * Sets a JSON object at the registry entry corresponding to the publicKey and dataKey.
 *
 * @param this - SkynetClient
 * @param privateKey - The user private key.
 * @param dataKey - The key of the data to fetch for the given user.
 * @param json - The JSON data to set.
 * @param [revision] - The revision number for the data entry.
 * @param [customOptions] - Additional settings that can optionally be set.
 * @throws - Will throw if the given entry revision does not fit in 64 bits, or if the revision was not given, if the latest revision of the entry is the maximum revision allowed.
 */


async function setJSON(privateKey, dataKey, json, revision, customOptions) {
  /* istanbul ignore next */
  if (typeof privateKey !== "string") {
    throw new Error("Expected parameter privateKey to be type string, was type ".concat(typeof privateKey));
  }
  /* istanbul ignore next */


  if (typeof dataKey !== "string") {
    throw new Error("Expected parameter dataKey to be type string, was type ".concat(typeof dataKey));
  }

  if (!(0, _string.isHexString)(privateKey)) {
    throw new Error("Expected parameter privateKey to be a hex-encoded string");
  }

  if (typeof json !== "object" || json === null) {
    throw new Error("Expected parameter json to be an object");
  }

  const opts = { ...this.customOptions,
    ...customOptions
  }; // Create the data to upload to acquire its skylink.

  const file = new File([JSON.stringify(json)], dataKey, {
    type: "application/json"
  }); // Start file upload, do not block.

  const skyfilePromise = this.uploadFile(file, opts);
  let skyfile;

  if (revision === undefined) {
    // fetch the current value to find out the revision.
    const {
      publicKey
    } = _tweetnacl.sign.keyPair.fromSecretKey((0, _string.hexToUint8Array)(privateKey)); // start getEntry, do not block.


    const entryPromise = this.registry.getEntry((0, _string.toHexString)(publicKey), dataKey, opts);
    let entry; // Block until both getEntry and Skyfile upload are finished.

    [entry, skyfile] = await Promise.all([entryPromise, skyfilePromise]);

    if (entry.entry === null) {
      revision = BigInt(0);
    } else {
      revision = entry.entry.revision + BigInt(1);
    } // Throw if the revision is already the maximum value.


    if (revision > _number.MAX_REVISION) {
      throw new Error("Current entry already has maximum allowed revision, could not update the entry");
    }
  } else {
    skyfile = await skyfilePromise;
  } // Assert the input is 64 bits.


  (0, _number.assertUint64)(revision); // build the registry value

  const entry = {
    datakey: dataKey,
    data: (0, _string.trimUriPrefix)(skyfile.skylink, _skylink.uriSkynetPrefix),
    revision
  }; // Update the registry.

  await this.registry.setEntry(privateKey, entry);
}
